design/0002-ci-health-metrics.md

```diff
@@ -0,0 +1,154 @@
+Annotations are surfaced through Github's Checks API.
+
+### Artifacts
+
+Instead of conforming to Github's annotation format, we could use a more
+standard format such as xUnit XML reports. That would imply building some
+wrapper around the tests to accommodate that format, and the results would be
+then uploaded to the CI run's artifact repository (which we're already
+leveraging to store docker images). Then another tool can consume those reports.
+
+But I haven't found such a tool that would consume xUnit reports for showing
+historical trends, which is the main goal of this RFC (being able to catch
+consistently flaky tests). For this reason I'm proposing we go down the
+annotations route.
+
+(Although do raise your voice if you find an appropriate tool for xUnit
+reports!)
+
+## Formatting
+
+We have integration tests and checks written in go, javascript and bash. Each of
+those environments would require a wrapper around their output to properly
+format errors as explained above. The error message would need to have at least
+two elements: A generic message that can be aggregated in a counter (see below)
+and a message with more details about this particular run.
+
+## Architecture
+
+The plan being to ultimately present the historical trend of errors, the obvious
+underlying data store would be a time series db.
+
+A familiar architecture could then be a Prometheus exporter fed off of data
+gathered from the Github API, collected by Prometheus and displayed by Grafana.
+
+### Prometheus exporter
+
+A process would poll the Github Actions and Checks API and expose a `/metrics`
+endpoint with the Prometheus metrics. Such metrics would be counters for each
+generic error message and for each workflow/job/step.
+
+This is the list of needed API calls:
+
+#### List all the failed CI runs in the repo
+
+``` GET /repos/linkerd/linkerd2/actions/runs?status=failed ``` (add the
+necessary paramters to only retrieve the runs since the last poll)
+
+This gives us a list of check suite IDs.
+
+#### List of check runs
+
+For each check suite ID, call
+
+``` GET /repos/linkerd/linkerd2/check-suites/:check_suite_id/check-runs ```
+
+Which returns each check run ID.
+
+#### List of annotations
+
+For each check run ID, we invoke the annotations API:
+
+``` GET repos/linkerd/linkerd2/check-runs/559037030/annotations ```
+
+That should give us the error message that we can use to build the metric.
+
+### Grafana
+
+The main end result can be as simple as a Bar Gauge graph representing the
+number of occurrences for each error message during the selected time span.
+
+These graphs could be aggregated per workflow, job and step.
+
+### Back-filling
+
+In order to start seeing results right away we need to back-fill Prometheus TSDB
+using all the results returned from the first API call referred above. Using
+[this approach](https://github.com/jjneely/stuff/tree/master/cmd/historygen)
+with some modifications I've been able to successfully achieve that.
```

Comment:
In order to do back-filling, we would need to generate annotations from the log output of previous runs, right?  How would that work?

Reply:
The Github API calls I described above will return all the historical CI runs data from the past few months (right now I don't know exactly how far back)

Reply:
Ah, but we'd only be able to backfill back to the point when we started formatting error messages as annotations, right?

Reply:
True, I missed that detail! Back-filling would only capture the generic "Process completed with exit code 1." messages, so this won't be worth it after all...